hybrid/ddc_compartment_switching/shared.s

```diff
@@ -0,0 +1,138 @@
+// Copyright (c) 2021 The CapableVMs "CHERI Examples" Contributors.
+// SPDX-License-Identifier: MIT OR Apache-2.0
+
+.global compartment_simple_fun
+
+.text
+.balign 4
+.global switch_compartment
+.type switch_compartment, "function"
+switch_compartment:
+    // int switch_compartment(int a, int b, void * stack, size_t size)
+    //
+    // For the purposes of this demo, `stack + size` must be 16-byte-aligned, so
+    // that it is suitable for use as a stack pointer with no additional
+    // alignment logic. In addition, the range [stack, stack+size) must be
+    // exactly representable as capability bounds.
+    //
+    // Note that, for this demo, the PCC is not restricted at all. Hybrid
+    // CheriBSD targets give the PCC very wide bounds, and all 'Load'
+    // permissions. In a real deployment, the PCC should be suitably restricted
+    // too, but this is likely to require co-operation with the C toolchain and
+    // runtime environment, since it will probably use the PCC to access literal
+    // pools, globals and other data outside the function itself.
+    //
+    // The new stack space must be wholly enclosed by the current DDC.
+    //
+    // The procedure-call standard is assumed to be AAPCS64 with the Morello
+    // supplement[1]. This is known as "hybrid" (as opposed to "purecap") mode.
+    //
+    // [1]: https://developer.arm.com/documentation/102205/latest
+    //
+    // Arguments are arranged as follows:
+    //
+    //      x0: stack   (pointer, not capability, since this is hybrid)
+    //      x1: size
+    //
+    // The result will be returned in w0.
+
+    // Derive a new DDC to cover the new stack.
+    mrs       c10, DDC
+    scvalue   c11, c10, x0
+    scbndse   c11, c11, x1
+    msr       DDC, c11
+
+    // Replace the stack pointer.
+    mov       x12, sp
+    add       sp, x0, x1
+
+    // Save the old DDC, stack pointer and return address on the new stack, so
+    // we can restore it when we return..
+    // This is the leaky part of the compartmentalisation. If strict
+    // compartments are required, some other technique must be used, such as a
+    // privileged switcher or sealing mechanism (e.g. using `ldpblr`).
+    str       c10, [sp, #-32]!
+    stp       x12, lr, [sp, #16]
+
+    // Stack layout at this point:
+    //
+    //     `stack + size` -> ________________________
+    //            sp + 24 -> [      old LR       ]   ^
+    //            sp + 16 -> [      old SP       ]   |
+    //            sp +  8 -> [ old DDC (high 64) ]   | DDC bounds
+    //            sp +  0 -> [ old DDC (low 64)  ]   |
+    //                                 :             :
+    //            `stack` -> ________________________v
+    //
+    // Note that this is _not_ an AAPCS64 frame record, even though it looks a
+    // bit like one. We don't touch FP here, and since it is not a capability
+    // (in hybrid mode), unwinding would fail anyway.
+
+    // Clean capabilities left in arguments.
+    
```

Comment:
This comment should go, now that we no longer have arguments to clean.

Reply:
Fix in 6284b24.